File: packages/server/src/routes/rooms.ts
import { Router } from "express";
import { eq, desc } from "drizzle-orm";
import { db } from "../db";
import { rooms, sessions, votes } from "../db/schema";
import { generateId, generateRoomCode } from "../lib/id";

const router = Router();

// POST /api/rooms — crear sala
router.post("/", (req, res) => {
  const { name } = req.body as { name?: string };
  if (!name || typeof name !== "string" || name.trim().length === 0) {
    res.status(400).json({ error: "name es requerido" });
    return;
  }

  const id = generateRoomCode();
  const now = Date.now();

  db.insert(rooms).values({ id, name: name.trim(), createdAt: now }).run();

  res.status(201).json({ id, name: name.trim(), createdAt: now });
});

// GET /api/rooms/:id — sala + sesión activa
router.get("/:id", (req, res) => {
  const room = db
    .select()
    .from(rooms)
    .where(eq(rooms.id, req.params.id.toUpperCase()))
    .get();

  if (!room) {
    res.status(404).json({ error: "Sala no encontrada" });
    return;
  }

  // Sesión activa = la más reciente sin result (no cerrada)
  const activeSession = db
    .select()
    .from(sessions)
    .where(eq(sessions.roomId, room.id))
    .orderBy(desc(sessions.createdAt))
    .limit(1)
    .get();

  const currentSession =
    activeSession && activeSession.result === null ? activeSession : null;

  // Votos de la sesión activa (si la hay y ya fue revelada)
  const currentVotes =
    currentSession && currentSession.revealedAt !== null
      ? db
          .select()
          .from(votes)
          .where(eq(votes.sessionId, currentSession.id))
          .all()
      : [];

  res.json({ room, currentSession, currentVotes });
});

// GET /api/rooms/:id/history — historial de sesiones cerradas
router.get("/:id/history", (req, res) => {
  const roomId = req.params.id.toUpperCase();

  const room = db.select().from(rooms).where(eq(rooms.id, roomId)).get();
  if (!room) {
    res.status(404).json({ error: "Sala no encontrada" });
    return;
  }

  const closedSessions = db
    .select()
    .from(sessions)
    .where(eq(sessions.roomId, roomId))
    .orderBy(desc(sessions.createdAt))
    .all()
    .filter((s) => s.result !== null);

  const history = closedSessions.map((session) => {
    const sessionVotes = db
      .select()
      .from(votes)
      .where(eq(votes.sessionId, session.id))
      .all();
    return { session, votes: sessionVotes };
  });

  res.json({ history });
});

export default router;

File: packages/server-python/routes/rooms.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from db import fetchone, fetchall, execute, generate_id, generate_room_code, row_to_camel, now_ms

router = APIRouter()


class CreateRoomBody(BaseModel):
    name: str


@router.post("", status_code=201)
async def create_room(body: CreateRoomBody):
    if not body.name.strip():
        raise HTTPException(status_code=400, detail="name es requerido")

    room_id = generate_room_code()
    ts = now_ms()
    name = body.name.strip()
    execute("INSERT INTO rooms (id, name, created_at) VALUES (?, ?, ?)", (room_id, name, ts))
    return {"id": room_id, "name": name, "createdAt": ts}


@router.get("/{room_id}")
async def get_room(room_id: str):
    room = fetchone("SELECT * FROM rooms WHERE id = ?", (room_id.upper(),))
    if not room:
        raise HTTPException(status_code=404, detail="Sala no encontrada")

    active = fetchone(
        "SELECT * FROM sessions WHERE room_id = ? ORDER BY created_at DESC LIMIT 1",
        (room["id"],),
    )
    current_session = active if active and active["result"] is None else None

    current_votes = []
    if current_session and current_session["revealed_at"] is not None:
        current_votes = fetchall(
            "SELECT * FROM votes WHERE session_id = ?", (current_session["id"],)
        )

    return {
        "room": row_to_camel(room),
        "currentSession": row_to_camel(current_session) if current_session else None,
        "currentVotes": [row_to_camel(v) for v in current_votes],
    }


@router.get("/{room_id}/history")
async def get_room_history(room_id: str):
    upper_id = room_id.upper()
    room = fetchone("SELECT * FROM rooms WHERE id = ?", (upper_id,))
    if not room:
        raise HTTPException(status_code=404, detail="Sala no encontrada")

    all_sessions = fetchall(
        "SELECT * FROM sessions WHERE room_id = ? ORDER BY created_at DESC", (upper_id,)
    )
    closed = [s for s in all_sessions if s["result"] is not None]

    history = []
    for session in closed:
        session_votes = fetchall("SELECT * FROM votes WHERE session_id = ?", (session["id"],))
        history.append({
            "session": row_to_camel(session),
            "votes": [row_to_camel(v) for v in session_votes],
        })

    return {"history": history}
